### PR TITLE
Change the stresslog message layout to support extra-large modules and modules loaded at far-apart addresses that use the stresslog

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Today, stresslog message strings that are more than 2^25 bytes after the start of the native module that contains the Stresslog code are recorded correctly. This leads to a few problems:

- If CoreCLR is too big (for example if it is instrumented by native sanitizers), the strings section can be too far away because of the extra code.
- If a user requests we load an out-of-module GC, it could be placed too far away.
- In NativeAOT, we don't control where the strings are placed and the runtime is linked into the same binary as the user's managed code. This can result in strings being placed out of range in large NativeAOT apps.

This change allows us to use format strings at any offset, at the cost of each message using more memory (so less messages recorded for the same allocated space).

Corresponding changes in the diagnostics repo are at dotnet/diagnostics#3779.

### CODE REVIEW

**Comment by Maoni0** [comment]
> @Maoni0 I'd like to know if reducing the effective size of the stress log by a little bit is a problem.

@PeterSolMS/@cshung, what do you think?

**Comment by jkoritzinsky** [comment]
> looking at your proposed data structure, I see that it is fairly well packed on 32bit platforms (your change adds 4 bytes to each stress log entry, moving them from having a 16byte header to having a 20byte header with 3 bytes of dead space), but inefficient on 64bit platforms (moving from a 16 byte header to a 32byte header with 11 bytes of dead space). Is there a reason you need the 64bit header to have the same layout as the 32bit header. On 64bit platforms if you placed the number of arguments BEFORE the format offset, you could have a 24byte header and only waste 3 bytes of space.

I believe we can have different layouts, as the diagnostic tools are built per-target. I’ll try that out.

**Comment by jkoritzinsky** [comment]
@dotnet/gc thoughts about the effective reduction in stresslog size due to the message size increase?

**Comment by cshung** [comment]
> @dotnet/gc thoughts about the effective reduction in stress log size due to the message size increase?

The GC team primarily uses the memory-mapped file to access the stress log. It is important that the `StressLogAnalyzer` stays functional. As a debugging tool only, it is not routinely tested in the CI, we need to manually check if it stays functional.

Regarding the size, right now we are mostly just redirecting all the `dprintf` statements to the stress log. This is currently using a lot of space, but there is a lot of room for optimization as well, we can simply dump the frequently logged and unnecessary strings.

@PeterSolMS for size concern - he claims to routinely use up 64GB of stress log space.

**Comment by jkoritzinsky** [comment]
Does StressLogAnalyzer need to be backward compatible with shipped stresslog formats, or can it version with the runtime's version? Right now it's set up to support the same format as the runtime (no back-compat detection). Backcompat wasn't preserved after the last change to the format so I didn't think it was worth it to build out support for it.

**Review by davidwrighton — APPROVED** [review]
Looks good to me.

**Comment by cshung** [comment]
> Does StressLogAnalyzer need to be backward compatible with shipped stresslog formats, or can it version with the runtime's version? Right now it's set up to support the same format as the runtime (no back-compat detection). Backcompat wasn't preserved after the last change to the format so I didn't think it was worth it to build out support for it.

I checked with @PeterSolMS about the backward compatibility - he is generally ok with that, but then he said he will think more about it. I will let him confirm.

**Comment by PeterSolMS** [comment]
I am indeed concerned about size. When debugging issues with server GC, we routinely use up a **lot** of stress log space, and there is nothing more annoying than finding that an interesting stress log message has probably been overwritten.

Of course you can always try to turn off irrelevant messages, but when debugging issues, it's hard to know ahead of time what's going to be relevant. And some of the issues we debug are hard to repro, so once you get the problem, you'd like to get to the bottom of it.

Bottom line is that I'd be very reluctant to spend additional space and thereby reduce stress log capacity. @cshung suggested to enable the change with #define in the cases where it matters. I'd be fine with that as long as the #define is off for the normal coreclr build.

**Comment by jkoritzinsky** [comment]
@PeterSolMS I'll give this another look to see if I can mess with the previous format to support larger binaries, but I'm not confident in having different message shapes depending on configuration (or even CoreCLR vs NativeAOT) to be a viable solution long term.

What do you think about us increasing the default stresslog size such that the default size per-thread and total size holds the same number of messages with the new format as the existing log holds with the current message layout?

**Comment by PeterSolMS** [comment]
@jkoritzinsky sorry about the delay in responding - increasing the default stress log size would do nothing for us - we never run with the default when debugging. I regularly set the total stress log size to something like 64 GB (half the physical memory on the machine), because it's always better to have more information available.

**Comment by jkoritzinsky** [comment]
@PeterSolMS if you set the size yourself, couldn't you increase the size you set it to so you'd continue to get about the same number of messages in the log?

**Comment by jkotas** [comment]
> @PeterSolMS if you set the size yourself, couldn't you increase the size you set it to so you'd continue to get about the same number of messages in the log?

You can do that only if your machine has enough physical memory for it. It is not unusual to use 100% of your machine memory when working on GC. You would have to buy extra memory to get equivalent results.

**Comment by jkotas** [comment]
Some comments on the format:
- If you are changing `formatOffset` to a pointer, you can just store the address of the message in the stress log instead of offset. There is extra logic in the stresslog logging to normalize the address into the offset, that was only necessary because of `formatOffset` played tricks with the bits. This will make stress log a bit faster.
- `timeStamp` is probably the least contentious place to steal the bits from. Today, the timeStamp covers 50,000 years if I am doing my math right, assuming typical timestamp granularity of 100ns. We can get away with a little bit less.

**Comment by cshung** [comment]
> > @PeterSolMS if you set the size yourself, couldn't you increase the size you set it to so you'd continue to get about the same number of messages in the log?
> 
> You can do that only if your machine has enough physical memory for it. It is not unusual to use 100% of your machine memory when working on GC. You would have to buy extra memory to get equivalent results.

We are using the memory-mapped files option - does that allow us to use more than physical memory?

**Comment by jkoritzinsky** [comment]
Based on feedback from @davidwrighton and @jkotas, I've changed the format to give us enough space for any reasonable scenario while keeping the message size the same (stealing bits from the timestamp until it can handle > 7 years, more than twice our LTS support timeframe for a given .NET version).

**Comment by jkoritzinsky** [comment]
@dotnet/gc, can I get another review pass? Now we aren't using any additional space, so stresslog capacity will stay the same.

File: src/coreclr/inc/stresslog.h [hunk 1]
```diff
@@ -544,25 +544,45 @@ inline BOOL StressLog::LogOn(unsigned facility, unsigned level)
...
-    uint64_t timeStamp;                     // time when mssg was logged
-    void*     args[0];                      // size given by numberOfArgs
+// The order of fields is important.  Ensure that we minimize padding
+// to fit more messages in a chunk.
+struct StressMsg
+{
+    static const size_t formatOffsetLowBits = 26;
+    static const size_t formatOffsetHighBits = 13;
+
+    uint32_t numberOfArgsLow  : 3;                   // at most 7 arguments here
```
Comment (jkotas) [inline src/coreclr/inc/stresslog.h:554]: This is getting pretty complicated. Do we really need to maintain the best-effort backward compatibility here?

Comment (jkoritzinsky) [reply src/coreclr/inc/stresslog.h:554]: We can shift around some of the structure, rev the version, and keep most of the complexity in SOS only for its back-compat.

Comment (jkotas) [reply src/coreclr/inc/stresslog.h:554]: I think it would be better - if the SOS side of the fix is not too complicated.

Comment (jkoritzinsky) [reply src/coreclr/inc/stresslog.h]: Updated the format to be cleaner and the corresponding change in the diagnostics repo to support reading both this format and the prior one.

File: src/coreclr/nativeaot/Runtime/inc/stressLog.h [hunk 2]
```diff
@@ -488,36 +488,89 @@ class StressLog {
...
+            uint64_t numberOfArgs : 6;                       // number of arguments
+            uint64_t formatOffsetLow: formatOffsetLowBits;   // offset of format string in modules
+            uint64_t formatOffsetHigh: formatOffsetHighBits; // offset of format string in modules
+            uint64_t timeStamp: 51;                          // time when msg was logged (100ns ticks since runtime start)
+        };
+
+        struct
+        {
+            uint64_t argsFacilityOffsetLow;
+            uint64_t timeStampOffsetHigh;
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]:
```suggestion
    struct
    {
        // We split the format offset to ensure that we utilize every bit and that
        // the compiler does not align the format offset to a new 64-bit boundary.
         uint64_t facility: 32;                           // facility used to log the entry
         uint64_t numberOfArgs : 6;                       // number of arguments
         uint64_t formatOffsetLow: formatOffsetLowBits;   // offset of format string in modules
         uint64_t formatOffsetHigh: formatOffsetHighBits; // offset of format string in modules
         uint64_t timeStamp: 51;                          // time when msg was logged (100ns ticks since runtime start)
```
Can this be just this? 

The union does not seem to be actually useful.

Comment (jkoritzinsky) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]: The union is required for the DebugHeader table (you can't do offsetof for bitfields)

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]: DebugHeader table describes the runtime data structure contracts. It works well for data structures that are composed from simple fields. It does not work well for bit fields.

We should not need offsets of these fields in the DebugHeader table since they are very imprecise descriptor of the contract.

Comment (jkoritzinsky) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]: If the debug header fields aren’t useful, I’ll remove the fields from the debug header and remove the union.

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]: Thanks. It will need a matching fix in the data contract reader (cc @mikem8361).

Comment (jkoritzinsky) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h:516]: I can set up the fix there as well if someone sends me the link to it on Teams.

Comment (jkoritzinsky) [reply src/coreclr/nativeaot/Runtime/inc/stressLog.h]: Updated the PR to remove the union.

**Review by jkotas — APPROVED** [review]
